container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/run_user_service_exec_commands.go

```
@@ -23,57 +24,110 @@ func RunUserServiceExecCommands(
+}
+
+func runExecOperationsInParallel(
+	ctx context.Context,
+	commandArgs map[service.ServiceUUID][]string,
+	userServiceDockerResources map[service.ServiceUUID]*shared_helpers.UserServiceDockerResources,
+	dockerManager *docker_manager.DockerManager,
+) (
+	map[service.ServiceUUID]*exec_result.ExecResult,
+	map[service.ServiceUUID]error,
+	error,
+) {
+	successfulExecs := map[service.ServiceUUID]*exec_result.ExecResult{}
+	failedExecs := map[service.ServiceUUID]error{}
 
-	// TODO Parallelize to increase perf
-	succesfulUserServiceExecResults := map[service.ServiceUUID]*exec_result.ExecResult{}
-	erroredUserServiceGuids := map[service.ServiceUUID]error{}
-	for guid, commandArgs := range userServiceCommands {
-		dockerResources, found := allDockerResources[guid]
+	execOperations := map[operation_parallelizer.OperationID]operation_parallelizer.Operation{}
+	for serviceUuid, commandArg := range commandArgs {
+		userServiceDockerResource, found := userServiceDockerResources[serviceUuid]
 		if !found {
-			erroredUserServiceGuids[guid] = stacktrace.NewError(
+			failedExecs[serviceUuid] = stacktrace.NewError(
 				"Cannot execute command '%+v' on service '%v' because no Docker resources were found for it",
 				commandArgs,
-				guid,
+				serviceUuid,
 			)
 			continue
 		}
-		container := dockerResources.ServiceContainer
 
+		execOperationId := operation_parallelizer.OperationID(serviceUuid)
+		execOperation := createExecOperation(ctx, serviceUuid, userServiceDockerResource, commandArg, dockerManager)
+		execOperations[execOperationId] = execOperation
+	}
+
+	successfulOperations, failedOperations := operation_parallelizer.RunOperationsInParallel(execOperations)
+	for operationUuid, operationResult := range successfulOperations {
+		serviceUuid := service.ServiceUUID(operationUuid)
+		execResult, ok := operationResult.(*exec_result.ExecResult)
+		if !ok {
+			return nil, nil, stacktrace.NewError("An error occurred processing the result of the exec command "+
```

Makes sense to exit here as this isn't a failed exec but something else completely.

---

Or should we add it to the `failedExecs` dictionary if there are others that might have succeeded?

---

I took some inspiration from what we do in add_services here because I was conflicted the same way you are.
I think failing hard is right here. Basically if you follow the codepath downstream, there's no real reason why one would fail for this reason and not the other (because the ExecResult object is computed the exact same way for all execs that a run in parallel). So, I believe that's the reasoning behind failing hard here, and it kind of made sense to me. I can add this as a comment in the code to make it clearer

---

I can live with that 👍 !